### ISSUE TITLE
BIRD VS DRONE CLASSIFICATION USING CNN

### ISSUE THREAD

**Issue by Jayamalar11** [Jayamalar11/feature-extracrion-on-mnist-dataset#1]
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.model_selection import train_test_split
import os
import shutil

!pip install -q kaggle

from google.colab import files
files.upload()  # This will prompt you to upload the kaggle.json file

!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json  # Set permissions for the file

!kaggle datasets download -d harshwalia/birds-vs-drone-dataset/data  # Example: !kaggle datasets download -d paultimothymooney/birds-vs-drones

!kaggle datasets download -d harshwalia/birds-vs-drone-dataset

import os
import shutil
from sklearn.model_selection import train_test_split

# Define paths to the original dataset
bird_path = '/content/drive/MyDrive/Birds'
drone_path = '/content/drive/MyDrive/Drones'

# Create directories for train, validation, and test sets
base_dir = '/content/split-dataset'
train_dir = os.path.join(base_dir, 'train')
val_dir = os.path.join(base_dir, 'val')
test_dir = os.path.join(base_dir, 'test')

for category in ['Birds', 'Drones']:
    os.makedirs(os.path.join(train_dir, category), exist_ok=True)
    os.makedirs(os.path.join(val_dir, category), exist_ok=True)
    os.makedirs(os.path.join(test_dir, category), exist_ok=True)

# Helper function to split and copy files
def split_and_copy_files(source_dir, train_dir, val_dir, test_dir, split_ratio=(0.7, 0.15, 0.15)):
    # Get list of all files in the source directory
    images = os.listdir(source_dir)

    # Split images into train, val, and test
    train_images, temp_images = train_test_split(images, test_size=(split_ratio[1] + split_ratio[2]), random_state=42)
    val_images, test_images = train_test_split(temp_images, test_size=split_ratio[2] / (split_ratio[1] + split_ratio[2]), random_state=42)

    # Copy images to their respective directories
    for img in train_images:
        shutil.copy(os.path.join(source_dir, img), os.path.join(train_dir, img))
    for img in val_images:
        shutil.copy(os.path.join(source_dir, img), os.path.join(val_dir, img))
    for img in test_images:
        shutil.copy(os.path.join(source_dir, img), os.path.join(test_dir, img))

# Split Birds and Drones
split_and_copy_files(bird_path, os.path.join(train_dir, 'Birds'), os.path.join(val_dir, 'Birds'), os.path.join(test_dir, 'Birds'))
split_and_copy_files(drone_path, os.path.join(train_dir, 'Drones'), os.path.join(val_dir, 'Drones'), os.path.join(test_dir, 'Drones'))

print("Dataset split into train, validation, and test sets.")

print(f"birds train images: {len(os.listdir(os.path.join(train_dir, 'Birds')))}")
print(f"birds val images: {len(os.listdir(os.path.join(val_dir, 'Birds')))}")

print(f"Drone train images: {len(os.listdir(os.path.join(train_dir, 'Drones')))}")
print(f"Drone val images: {len(os.listdir(os.path.join(val_dir, 'Drones')))}")

print(f"birds test images: {len(os.listdir(os.path.join(test_dir, 'Birds')))}")
print(f"Drone test images: {len(os.listdir(os.path.join(test_dir, 'Drones')))}")

from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Define augmentation for training
train_datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest'
)

# No augmentation for validation and test, just rescaling
val_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)

# Load images and apply augmentation
train_generator = train_datagen.flow_from_directory(
    train_dir,
    target_size=(100,100),  # Resize images if needed
    batch_size=32,
    class_mode='categorical'  # Use 'binary' if only 2 classes
)

val_generator = val_datagen.flow_from_directory(
    val_dir,
    target_size=(100,100),
    batch_size=32,
    class_mode='categorical'
)

test_generator = test_datagen.flow_from_directory(
    test_dir,
    target_size=(100,100),
    batch_size=32,
    class_mode='categorical',
    shuffle=False  # Important for generating predictions later
)

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

# Create a CNN model
model = Sequential()

# First convolutional block
model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(100,100, 3)))
model.add(MaxPooling2D((2, 2)))

# Second convolutional block (added)
model.add(Conv2D(64, (3, 3), activation='relu'))  # New conv layer
model.add(MaxPooling2D((2, 2)))

# Third convolutional block
model.add(Conv2D(128, (3, 3), activation='relu'))
model.add(MaxPooling2D((2, 2)))

# Fourth convolutional block (added)
model.add(Conv2D(256, (3, 3), activation='relu'))  # New conv layer
model.add(MaxPooling2D((2, 2)))

# Flatten and fully connected layers
model.add(Flatten())
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.5))  # Dropout to reduce overfitting
model.add(Dense(2, activation='sigmoid'))  # 2 classes: bird and drone

# Summarize the model
model.summary()

from tensorflow.keras.optimizers import Adam

# Define the learning rate
lr = 0.0001

# Compile the model
model.compile(loss='binary_crossentropy',  # Use 'binary_crossentropy' for binary classification
              optimizer=Adam(learning_rate = lr),
              metrics=['accuracy'])

# Train the model
steps_per_epoch = train_generator.samples // train_generator.batch_size

history = model.fit(
    train_generator,
    steps_per_epoch= steps_per_epoch,
    validation_data=val_generator,
    validation_steps=val_generator.samples // val_generator.batch_size,
    epochs=50
    )

# Evaluate the model on the test data
test_loss, test_acc = model.evaluate(test_generator, steps=test_generator.samples // test_generator.batch_size)
print(f'Test accuracy: {test_acc:.2f}')

import numpy as np
# Ensure that the number of steps is correctly set
steps = test_generator.samples // test_generator.batch_size
if test_generator.samples % test_generator.batch_size != 0:
    steps += 1  # Ensure the last partial batch is included

# Generate predictions on the test set
predictions = model.predict(test_generator, steps=steps)

# Convert predicted probabilities to class indices
predicted_classes = np.argmax(predictions, axis=1)

# Get the true class labels from the test generator
true_classes = test_generator.classes

# Ensure that the length of true and predicted classes is consistent
assert len(true_classes) == len(predicted_classes), f"Mismatch in number of samples: {len(true_classes)} vs {len(predicted_classes)}"

# Confusion Matrix
from sklearn.metrics import confusion_matrix
conf_matrix = confusion_matrix(true_classes, predicted_classes)
print(f'Confusion Matrix:\n{conf_matrix}')

import matplotlib.pyplot as plt
import numpy as np
from random import sample

# Define class labels
class_labels = ['Birds', 'Drones']

# ✅ Function to visualize images from each category
def visualize_images(indices, title, num_samples=10):
    indices = sample(list(indices), min(num_samples, len(indices)))  # Select 10 random samples (or available)
    plt.figure(figsize=(15, 5))
    for i, idx in enumerate(indices):
        plt.subplot(2, 5, i+1)
        plt.imshow(test_images[idx].astype(np.uint8))  # Ensure images are in uint8 format for proper display
        plt.axis('off')
        plt.title(f"True: {class_labels[true_labels[idx]]}\nPred: {class_labels[pred_labels[idx]]}")
    plt.suptitle(title, fontsize=16)
    plt.show()

# ✅ Display images for each category
visualize_images(TP_indices, "True Positives (Correctly Classified)", num_samples=10)
visualize_images(TN_indices, "True Negatives (Correctly Rejected)", num_samples=10)
visualize_images(FP_indices, "False Positives (Incorrectly Classified as Positive)", num_samples=10)
visualize_images(FN_indices, "False Negatives (Incorrectly Classified as Negative)", num_samples=10)

import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, auc
from sklearn.preprocessing import label_binarize
from itertools import cycle

# Assuming true_classes is defined elsewhere and contains the true labels
# Binarize the true classes for binary classification
n_classes = len(test_generator.class_indices)
y_test_bin = label_binarize(true_classes, classes=[0, 1])  # Specify classes for binary

# Check the shape of predictions
print("Shape of predictions:", predictions.shape)
print("Shape of y_test_bin:", y_test_bin.shape)

# Check the unique classes in true_classes
unique_classes = np.unique(true_classes)
print("Unique classes in true_classes:", unique_classes)

# If there are not two unique classes, skip ROC calculation
if len(unique_classes) < 2:
    print("Warning: Insufficient unique classes for ROC curve computation. Need at least two classes.")
else:
    # Calculate ROC curve and AUC for binary classification
    fpr, tpr, _ = roc_curve(y_test_bin.ravel(), predictions[:, 1])  # Use predictions for the positive class
    roc_auc = auc(fpr, tpr)

    # Plot ROC curve
    plt.figure(figsize=(10, 8))
    plt.plot(fpr, tpr, color='darkorange', lw=2,
             label=f'ROC curve (area = {roc_auc:0.2f}) for positive class')
    plt.plot([0, 1], [0, 1], 'k--', lw=2)
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver Operating Characteristic (ROC) Curve')
    plt.legend(loc="lower right")
    plt.show()

from tensorflow.keras.preprocessing import image
import numpy as np
import matplotlib.pyplot as plt

# Class labels
class_labels = {0: 'Bird', 1: 'Drone'}  # Mapping the class index to labels

# Function to classify a single image
def classify_image(img_path, model):
    # Load the image with target size (must match the input shape of the model)
    img = image.load_img(img_path, target_size=(100,100))  # Resize to the model's input size

    # Convert the image to a NumPy array and preprocess it
    img_array = image.img_to_array(img)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension
    img_array /= 255.0  # Rescale the pixel values to [0, 1] (same as during training)

    # Predict the class (Bird or Drone)
    predictions = model.predict(img_array)

    # Get the predicted class index
    predicted_class = np.argmax(predictions, axis=1)[0]  # 0 is for bird, 1 is for drone

    # Get the predicted class label
    class_label = class_labels[predicted_class]

    # Print and return the class label and probability
    print(f"Predicted class: {class_label} (Probability: {predictions[0][predicted_class]:.2f})")

    # Display the image
    img = image.load_img(img_path)
    plt.imshow(img)
    plt.title(f"Predicted: {class_label}")
    plt.axis('off')
    plt.show()

# Example usage
img_path = '/content/drive/MyDrive/test_b1.jpeg'  # Provide the path to the image you want to classify
classify_image(img_path, model)
